kernel/src/predicates/mod.rs

```diff
@@ -0,0 +1,652 @@
+use crate::expressions::{
+    BinaryOperator, ColumnName, Expression as Expr, Scalar, UnaryOperator, VariadicOperator,
+};
+use crate::schema::DataType;
+
+use std::cmp::Ordering;
+use tracing::debug;
+
+#[cfg(test)]
+mod tests;
+
+/// Evaluates a predicate expression tree against column names that resolve as scalars. Useful for
+/// testing/debugging but also serves as a reference implementation that documents the expression
+/// semantics that kernel relies on for data skipping.
+///
+/// # Inverted expression semantics
+///
+/// Because inversion(`NOT` operator) has special semantics and can often be optimized away by
+/// pushing it down, most methods take an `inverted` flag. That allows operations like
+/// [`UnaryOperator::Not`] to simply evaluate their operand with a flipped `inverted` flag,
+///
+/// # NULL and error semantics
+///
+/// Literal NULL values almost always produce cascading changes in the predicate's structure, so we
+/// represent them by `Option::None` rather than `Scalar::Null`. This allows e.g. `A < NULL` to be
+/// rewritten as `NULL`, or `AND(NULL, FALSE)` to be rewritten as `FALSE`.
+///
+/// With the exception of `IS [NOT] NULL`, all operations produce NULL output if any input is
+/// `NULL`. Any resolution failures also produce NULL (such as missing columns or type mismatch
+/// between a column and the scalar it is compared against).
+///
+/// For safety reasons, this evaluator only accepts expressions of the form `<col> IS [NOT]
+/// NULL`. This is because stats-based skipping is only well-defined for that restricted case --
+/// there is no easy way to distinguish whether NULL was due to missing stats vs. an operation that
+/// produced NULL for other reasons.
+///
+/// NOTE: The error-handling semantics of this trait's scalar-based predicate evaluation may differ
+/// from those of the engine's expression evaluation, because kernel expressions don't include the
+/// necessary type information to reliably detect all type errors.
+pub(crate) trait PredicateEvaluator {
+    type Output;
+
+    /// A (possibly inverted) boolean scalar value, e.g. `[NOT] <value>`.
+    fn eval_scalar(&self, val: &Scalar, inverted: bool) -> Option<Self::Output>;
+
+    /// A (possibly inverted) NULL check, e.g. `<expr> IS [NOT] NULL`.
+    fn eval_is_null(&self, col: &ColumnName, inverted: bool) -> Option<Self::Output>;
+
+    /// A less-than comparison, e.g. `<col> < <value>`.
+    ///
+    /// NOTE: Caller is responsible to commute and/or invert the operation if needed,
+    /// e.g. `NOT(<value> < <col>` becomes `<col> <= <value>`.
+    fn eval_lt(&self, col: &ColumnName, val: &Scalar) -> Option<Self::Output>;
+
+    /// A less-than-or-equal comparison, e.g. `<col> <= <value>`
+    ///
+    /// NOTE: Caller is responsible to commute and/or invert the operation if needed,
+    /// e.g. `NOT(<value> <= <col>` becomes `<col> > <value>`.
```

Comment:
Commuting (reordering) is needed because the eval logic that follows is based on `col` being on the left and `value` being on the right. Otherwise there are twice as many cases to implement.

Reply:
ahhh got it :) thanks!

Reply:
disregard then! (though i think the top one still applies)

Reply:
Nice catch on the `>` vs. `<` btw!